### PR TITLE
Add imm data APIs.

### PR DESCRIPTION
**Pull request by GTwhy**

### CODE REVIEW

File: src/agent.rs [hunk 1]
```diff
@@ -262,11 +282,21 @@ impl AgentThread {
...
-            let sz = self
+            for i in header_buf.as_mut_slice() {
+                *i = 0_u8;
+            }
+            let (sz, imm) = self
                 .inner
                 .qp
                 .receive_sge(&[&mut header_buf, &mut data_buf])
                 .await?;
+            if header_buf.as_slice() == CLEAN_STATE && imm.is_some() {
```
Comment (rogercloud) [inline src/agent.rs:293]: exchange these two condition, for non-imm request we don't need to compare slice.

File: src/lib.rs [hunk 2]
```diff
@@ -432,7 +432,100 @@ impl Rdma {
         self.agent
             .as_ref()
             .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "Agent is not ready"))?
-            .send_data(lm)
+            .send_data(lm, None)
```
Comment (rogercloud) [inline src/lib.rs:435]: reuse `send_with_imm` function?

File: src/lib.rs [hunk 3]
```diff
@@ -483,13 +576,179 @@ impl Rdma {
     /// ```
     #[inline]
     pub async fn receive(&self) -> io::Result<LocalMr> {
+        let (lmr, _) = self
```
Comment (rogercloud) [inline src/lib.rs:579]: reuse `receive_with_imm` function?

File: src/lib.rs (hunk 2 above)
Comment (GTwhy) [reply src/lib.rs:435]: `send_with_imm`  forces the user to fill in a u32, so reuse is not appropriate here.

File: src/agent.rs [hunk 4]
```diff
@@ -65,9 +67,11 @@ impl Agent {
         let (local_mr_send, local_mr_recv) = channel(1024);
         let (remote_mr_send, remote_mr_recv) = channel(1024);
         let (data_send, data_recv) = channel(1024);
+        let (imm_send, imm_recv) = channel(1024);
         let local_mr_recv = Mutex::new(local_mr_recv);
         let remote_mr_recv = Mutex::new(remote_mr_recv);
-        let data_recv_mutex = Mutex::new(data_recv);
+        let data_recv = Mutex::new(data_recv);
```
[the remark this answers is not in the document]
Comment (GTwhy) [reply src/agent.rs:73]: `Rdma` may be used by `spawn` so these fields use `Mutex`

[the remark this answers is not in the document]
Comment (rogercloud) [reply src/agent.rs:75]: tokio spawn only requires `Send` trait not `Sync`

**Review by rogercloud — APPROVED** [review]

**Pull request merged**